platform/linux-generic/odp_traffic_mngr.c

```diff
@@ -4438,6 +4445,37 @@ int odp_tm_enq_with_cnt(odp_tm_queue_t tm_queue, odp_packet_t pkt)
 	return pkt_cnt;
 }
 
+int odp_tm_enq_multi(odp_tm_queue_t tm_queue, const odp_packet_t packets[],
+		     int num)
+{
+	tm_queue_obj_t *tm_queue_obj;
+	tm_system_t *tm_system;
+	int i, rc;
+
+	tm_queue_obj = GET_TM_QUEUE_OBJ(tm_queue);
+	if (!tm_queue_obj)
+		return -1;
+
+	tm_system = &tm_glb->system[tm_queue_obj->tm_idx];
+	if (!tm_system)
+		return -1;
+
+	if (odp_atomic_load_u64(&tm_system->destroying))
+		return -1;
+
+	for (i = 0; i < num; i++) {
+		rc = tm_enqueue(tm_system, tm_queue_obj, packets[i]);
+		if (rc < 0)
+			break;
+	}
+
+	/* Free rejected buffers */
+	for (; i < num; i++)
+		odp_packet_free(packets[i]);
```

Comment:
@psavol noticed that there is something wrong here. An indeed there is. If tm_enqueue() decides to drop a packet because of WRED, it does not mean that all subsequent packets would be dropped. So freeing the rest of the packets in this loop is likely to drop packets that would go through without the probability based WRED algorithm dropping them. In other words, this effectively breaks the logic of the WRED dropper.

In case tm_enqueue() returns a failure because of WRED drop, the correct thing to do would be to free only that packet and to continue calling tm_enqueue() for the rest of the packets.

One problem is that tm_enqueue() does not tell if it is dropping a packet because of WRED or for other reasons. The simple thing to do would be to continue as if all failures were because of WRED (which is likely). Another possibility  would be to distinguish WRED drops from other failures in the return code of tm_enqueue(). It seems that in case of WRED drop, tm_enqueue() could easily return e.g. -2 instead of -1 and tm_enq_multi() could then check that. In case of other tm_enqueue() failures than WRED, tm_enq_multi() should stop calling tm_enqueue() and then return the number of packets consumed that far without freeing the rest.

Reply:
Ack, fixed it by changing return value of red drop to -2 and freeing pkt and continuing if it is red drop

Reply:
Now that WRED drops are distinguished from other errors, the function should not consume all packets if it encounters an other error than WRED drop. So instead of freeing the rest of the packets and returning num, the function should not free the packets but return the number of packets consumed before the error (i.e. the value of i).

This also means that tail drop (in addition to some real internal errors) does not cause packets to be consumed but the application must do the actual drop in that case. In some cases that might be slightly inconvenient but it allows an application to do something else than drop the remaining packets if there is no RED configured and the queue becomes full.